modules/fedimint-meta-client/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+pub mod api;
+#[cfg(feature = "cli")]
+pub mod cli;
+pub mod db;
+pub mod states;
+
+use std::collections::BTreeMap;
+
+use api::MetaFederationApi;
+use common::{MetaConsensusValue, MetaKey, MetaValue};
+use db::DbKeyPrefix;
+use fedimint_client::db::ClientMigrationFn;
+use fedimint_client::module::init::{ClientModuleInit, ClientModuleInitArgs};
+use fedimint_client::module::recovery::NoModuleBackup;
+use fedimint_client::module::{ClientModule, IClientModule};
+use fedimint_client::sm::Context;
+use fedimint_core::api::DynModuleApi;
+use fedimint_core::core::Decoder;
+use fedimint_core::db::{DatabaseTransaction, DatabaseVersion};
+use fedimint_core::module::{
+    ApiAuth, ApiVersion, ModuleCommon, ModuleInit, MultiApiVersion, TransactionItemAmount,
+};
+use fedimint_core::{apply, async_trait_maybe_send, Amount, PeerId};
+pub use fedimint_meta_common as common;
+use fedimint_meta_common::{MetaCommonInit, MetaModuleTypes};
+use states::MetaStateMachine;
+use strum::IntoEnumIterator;
+
+#[derive(Debug)]
+pub struct MetaClientModule {
+    module_api: DynModuleApi,
+    admin_auth: Option<ApiAuth>,
```

Comment:
future: Should this be given via some context struct given to `handle_cli_command`?

Reply:
It is given via `init` or `recovery`, so no need pass it via `handle_cli_command`, IMO.

Reply:
But should it? That way every module will have to keep its own copy. Not a big deal though either way.

Reply:
Only modules that need it will keep a copy...

But I got confused before and I thought we're talking about the `module_api` and `global_api`, which are now different (have only one peer in them) when the client is "guardian client (`--our-id` passed)".

I guess is now an architectural question now:

Is the whole `Client` started in a "guardian mode", or is it just a single command? Because in `fedimint-cli` we build a whole client for a single command every time, so these are currently conflated and hard to tell.

E.g. would it make sense to have "guardian only State Machines"? So guardians could kickstart longer running things on their clients that would make guardian-only API calls?

@elsirion 

I kind of thing it makes sense to support it eventually, but didn't made any drastic moves yet.

Also there's still the core question from https://github.com/fedimint/fedimint/issues/4519 : if you're a `Client` started by the guardian, should we even bother calling other guardians if you can fully trust your own? Maybe for redundancy ... ? In that case we'll need to have `all_peers` + `guardian_peer` and on the traits, and use one or the other on appropriate calls.

Reply:
> E.g. would it make sense to have "guardian only State Machines"? So guardians could kickstart longer running things on their clients that would make guardian-only API calls?

Could be useful eventually, but low priority imo.

> Also there's still the core question from https://github.com/fedimint/fedimint/issues/4519 : if you're a Client started by the guardian, should we even bother calling other guardians if you can fully trust your own? Maybe for redundancy ... ? In that case we'll need to have all_peers + guardian_peer and on the traits, and use one or the other on appropriate calls.

As it stands right now we have to only use the guardian peer if authentication is present because we leak our secret otherwise? Ideally we could tell the admin functions to only call the guardian API while all others act normally. There might be cases where a guardian does maintenaince but still wants to be able to transact.

Reply:
> As it stands right now we have to only use the guardian peer if authentication is present because we leak our secret otherwise?

We only send auth with rpcs that require it, and these use separate methods.

>  There might be cases where a guardian does maintenaince but still wants to be able to transact.


Yeah, so we'll need the trait to have separate `all_peers` for normal calls and `our_peer_id() -> Option<PeerId>` for own guardian APIs. Small refactor, I believe.

Reply:
> Could be useful eventually, but low priority imo.

We don't have to do it. Module authors probabl already can because they get everything in `init`&`recovery` args.